nx/lib/nx.ex

```diff
@@ -5176,15 +5176,114 @@ defmodule Nx do
+      impl!(target, indices, updates).indexed_add(out, target, indices, updates)
+    end
+
+    indexed_op(target, indices, updates, impl)
+  end
+
+  @doc """
+  Performs an indexed `set` operation on the `target` tensor,
+  setting the `updates` into the corresponding `indices` positions.
+
+  `indices` must be a fully qualified tensor of shape `{n, Nx.rank(target)}`, with `n`
+  being an arbitrary number of indices, while `updates` must have a compatible `{n}` shape.
+
+  In case of repeating indices, the last occurence of index and its corresponding update
+  value wins.
+
+  ### Examples
+
+      iex> Nx.indexed_put(Nx.tensor([0, 0, 0]), Nx.tensor([[1], [2]]), Nx.tensor([2, 4]))
+      #Nx.Tensor<
+        s64[3]
+        [0, 2, 4]
+      >
+
+      iex> Nx.indexed_put(Nx.tensor([0, 0, 0]), Nx.tensor([[1], [2], [1]]), Nx.tensor([3, 4, 2]))
+      #Nx.Tensor<
+        s64[3]
+        [0, 2, 4]
+      >
+
+      iex> t = Nx.iota({1, 2, 3})
+      #Nx.Tensor<
+        s64[1][2][3]
+        [
+          [
+            [0, 1, 2],
+            [3, 4, 5]
+          ]
+        ]
+      >
+      iex> indices = Nx.tensor([[0, 0, 0], [0, 1, 1], [0, 0, 0], [0, 0, 2]])
+      iex> updates = Nx.tensor([1, 3, 2, -2])
+      iex> Nx.indexed_put(t, indices, updates)
+      #Nx.Tensor<
+        s64[1][2][3]
+        [
+          [
+            [2, 1, -2],
+            [3, 3, 5]
+          ]
+        ]
+      >
+
+  Type promotions should happen automatically.
+
+      iex> Nx.indexed_put(Nx.tensor([1.0]), Nx.tensor([[0]]), Nx.tensor([3]))
+      #Nx.Tensor<
+        f32[1]
+        [3.0]
+      >
+
+      iex> Nx.indexed_put(Nx.tensor([1]), Nx.tensor([[0]]), Nx.tensor([3.0]))
+      #Nx.Tensor<
+        f32[1]
+        [3.0]
+      >
+
+      iex> Nx.indexed_put(Nx.tensor([1], type: {:s, 32}), Nx.tensor([[0]]), Nx.tensor([3], type: {:s, 64}))
+      #Nx.Tensor<
+        s64[1]
+        [3]
+      >
+
+  ### Error cases
+      iex> Nx.indexed_put(Nx.tensor([[1], [2]]), Nx.tensor([[[1, 2, 3]]]), Nx.tensor([0]))
+      ** (ArgumentError) indices must be a rank 2 tensor, got: 3
+
+      iex> Nx.indexed_put(Nx.tensor([[1], [2]]), Nx.tensor([[1, 2]]), Nx.tensor([[0]]))
+      ** (ArgumentError) updates must be a rank 1 tensor, got: 2
+
+      iex> Nx.indexed_put(Nx.tensor([[1], [2]]), Nx.tensor([[1, 2, 3]]), Nx.tensor([0]))
+      ** (ArgumentError) expected indices to have shape {*, 2}, got: {1, 3}
+
+      iex> Nx.indexed_put(Nx.tensor([[1], [2]]), Nx.tensor([[1, 2]]), Nx.tensor([0, 1]))
+      ** (ArgumentError) expected updates tensor to match the first axis of indices tensor with shape {1, 2}, got {2}
+  """
+  @doc type: :indexed
+  def indexed_put(target, indices, updates) do
+    impl = fn target, indices, updates, out ->
+      impl!(target, indices, updates).indexed_put(out, target, indices, updates)
+    end
+
+    indexed_op(target, indices, updates, impl)
+  end
+
+  defp indexed_op(target, indices, updates, impl) do
```

Comment:
```suggestion
    indexed_op(target, indices, updates, :indexed_put)
  end

  defp indexed_op(target, indices, updates, function) do
```

Reply:
This and the suggestion below seems better to avoid floating anonymous functions around. Also makes things clearer in the sense that the implicit contract for the anonymous function had `out` as the last argument instead of the usual first one